### DISCUSSION TITLE
Optimistic updates across multiple query keys

### DISCUSSION THREAD

**Discussion by ianduvall** [TanStack/query#1780] [Q&A]
I'm displaying a list of items that can be sorted by items' attributes. This leads to multiple cache keys for different sort values. When making mutations to items, e.g. change the name of the item, my backend doesn't synchronously update the items with those changes (it can take a few seconds) so I need to optimistically update the items across all of the different cache keys. I've tried `setQueryData` but that doesn't seem to support query filters. Any other ideas on how to achieve this?

**Comment by TkDodo** [comment]
I usually only optimistically update the one view that the user currently sees, and solve the rest with query invalidation, which is fuzzy.

**Reply by babycourageous** [reply]
I'm curious - How do you know what view to update? Are you using a different mutation hook for each view? 

I have a use case where I'm optimistically updating all queries affected by a mutation cause I'm not sure what view the user might be on. 
I'm currently using a react router hook in my custom query hook to check for a potential URL parameter but still having to `if/else` or `switch/case` all possible queries and write all the different `setQueryData` logic out.

this is some super pseudo code to hopefully get the example across: 

```js
function useUpdateThing() {
  const params = useParams() // if on specific page there will be a "key3ID" param
  return useMutation ((variables) => mutateFn, 
  {
    onMutate: (variables) => {
        const view1PrevData =  queryClient.getQueryData('key1')
        const view2PrevData = queryClient.getQueryData('key2')
        const view3PrevData = queryClient.getQueryData(['key3', params.key3ID])

        if (view1) {
          // view1 setQueryData here
        }

        if (view2) {
          // view2 setQueryData here    
       }

        if (view3) {
          // view3 setQueryData here
        }

       // return fallback objects
       return { view1PrevData, view2PrevData, view3PrevData }
    },
    // etc
  })
}
```

hope that makes sense.

**Reply by TkDodo** [reply]
> I'm curious - How do you know what view to update? 

I've just double checked with our source code. We call `useMutation` in the same component that calls `useQuery`, so we have all filters available.

If that's not the case, I would store that information likely in the url (RouterState with react-router) . It depends on what it is obviously, but if I'm thinking about a list of todos that I can filter, where the keys are like:

```
['todos', { filter: 'all '}]
['todos', { filter: 'open' }]
['todos', { filter: 'done' }]
```

my url would be:

```
/todos/all/
/todos/open/
/todos/done/
```

every time the user changes their selection, all I have to do is update the url - no (other) local state necessary. It will automatically re-render all subscribers, so the useQuery changes to the correct view. When I do `useMutation` deeper down the tree, I can also just do `useParams` and compute the key from that. Abstracting that into a custom `useTodosQueryKey` hook that creates the queryKey based on the current router state sounds also nice.

But yes, it currently requires to be aware of all inputs to the queryKey, which does require some sort of "distributing" that state on the client. We also have one app where we use `redux` for that local state and it works fine.

**Reply by babycourageous** [reply]
Thanks so much! That makes sense. 

> But yes, it currently requires to be aware of all inputs to the queryKey, which does require some sort of "distributing" that state on the client. We also have one app where we use redux for that local state and it works fine.

Are you saying that the pseudo code makes sense (is the way to go about it) - where you go thru each cache entry you pulled and `setQueryData` for each? 

I would imagine so - especially since each cache entry has a different structure which means different ways to set it.

**Reply by TkDodo** [reply]
>  especially since each cache entry has a different structure which means different ways to set it.

if it all had the same structure and you could build the key reliably because every part is available, it would just be one line:

```
function useTodoMutation() {
  const queryKey = useTodosQueryKey() // build key from either url parameters or global state that knows the current selection / view
  return useMutation ((variables) => mutateFn, 
  {
    onMutate: (variables) => {
        const previousData =  queryClient.getQueryData<Array<Todo>>(queryKey)
        if (previousData) queryClient.setQueryData<Array<Todo>>(queryKey, ...)

       // return fallback objects
       return { previousData }
    },
    // etc
  })
}
```

This is where a "fuzzy" `setQueriesData(['todos'])` would _also_ come in handy, because you would not need to know the "exact" view you are on - it's all just `<Array<Todo>>`.

But if all views have different structures, I don't see how you could generically update it with one statement

**Reply by babycourageous** [reply]
>But if all views have different structures, I don't see how you could generically update it with one statement

100%. I always wonder if I'm missing something when i hit a more complicated use case. Glad to see I wasn't!

I can see now where a fuzzy `setQueriesData` might be handy for those cases where you have 4 cache entries that have the exact same structure throughout the app. Would be very handy indeed.

But, lucky for me mine all have different structures 😂... One's a detail, one is an array of details, another is a different thing altogether that has a detail as a piece of it. I have quite the interesting Optimistic Update heheh.

**Reply by TkDodo** [reply]
> I can see now where a fuzzy setQueriesData might be handy for those cases where you have 4 cache entries that have the exact same structure throughout the app. Would be very handy indeed.

yes, that would be the primary use-case. It should be as simple as:

```
notifyManager.batch(() => {
  queryCache.findAll(filters).forEach(query => {
      query.setData(updater, options)
    })
  })
}
```

maybe you can try it out in user-land or in a sandbox and then we can make a PR to the library?

> But, lucky for me mine all have different structures 😂... One's a detail, one is an array of details, another is a different thing altogether that has a detail as a piece of it. I have quite the interesting Optimistic Update heheh.

I mean it could still work if you use the functional updater form, and then do some checking at runtime, like, if we had that function:

```
queryClient.setQueriesData(['todos], previous => {
  Array.isArray(previous) ? someArrayOption : somethingWithTheDetailsEntry
})
```

But yeah, if it's complex, it's complex . And TypeScript probably won't like that very much 😅

**Reply by babycourageous** [reply]
Yeah - it's complex but I think makes sense that for a case like mine that there wouldn't be a catch-all way to update - especially when TS gets involved, heheh. Just too many variables/unknowns at play. 

If I get some down time this week i would love to poke around the code. I'll let ya know!

**Reply by Myzel394** [reply]
> I usually only optimistically update the one view that the user currently sees, and solve the rest with query invalidation, which is fuzzy.

How do you fuzzily invalidate the other queries @TkDodo ?

**Reply by TkDodo** [reply]
it depends on your query key structure. Given:

```
['todos', { filter: 'all '}]
['todos', { filter: 'open' }]
['todos', { filter: 'done' }]
```

you can call `queryClient.invalidateQueries(['todos'])` to target all of them. if you want to exclude the view you are on, you can try to use the `predicate` filter function.

/edit sorry, even easier: invalidate the inactive ones only:

```
queryClient.invalidateQueries(['todos'], { type: 'inactive' })
```

**Reply by Myzel394** [reply]
@TkDodo Thank you for your answer. I finally figured out, I was using the wrong function, hence the late answer 😅 . One last question: Should I call `cancelQueries` before calling `invalidateQueries`? Also, can I use `setQueryData` fuzzily?

**Reply by TkDodo** [reply]
> Should I call cancelQueries before calling invalidateQueries

happens automatically in v4. In v3, you can pass `cancelRefetch` to `invalidateQueries` to re-start any running fetches.

> Also, can I use setQueryData fuzzily?

Yes, but if you want to target "all" matching keys, use `setQueriesData`

**Comment by phaux** [comment]
I just encountered the same problem.

I have a `useQuery(["things"])` but it can also be `useQuery(["things", {sortField, sortDirection}])`. I have a delete mutation which optimistically removes item from the cache, but it only updates in places which don't use the sorting.

It would be nice if we had something like `setQueryData(key, updater, { strict: false })`

**Reply by TkDodo** [reply]
> It would be nice if we had something like setQueryData(key, updater, { strict: false })

Maybe something like `setQueriesData` could be fuzzy similar to `invalidateQueries`. Recently discussed this behavior with @boschni here: https://github.com/tannerlinsley/react-query/issues/1725#issuecomment-773347647

The functional updater would need to be called for every entry, but if they have a different data structure, it will be quite hard to make that typesafe. What don you think @tannerlinsley ?

**Reply by phaux** [reply]
> but if they have a different data structure, it will be quite hard to make that typesafe.

It would be the programmer's job to filter only one type of queries using predicate function:

```ts
// optimistically remove todo
queryCache.setQueryData<Todo[]>(
  ["todos"],
  (todos) => todos.filter((todo) => todo.id != removedTodoId),
  {
    strict: false,
    predicate: ({ queryKey }) => queryKey[1] == null || typeof queryKey[1] == "object",
  }
)

**Comment by henrikq** [comment]
I ran into the same issue, and seemed solved it with a hook as follows:
```
function useUpdateQueryData(queryKey, updater) {
  const client = useQueryClient();
  return newData => {
    const cache = client.getQueryCache();
    cache.findAll(queryKey)
      .forEach(query => 
        client.setQueryData(query.queryKey, oldData => updater(oldData, newData))
      );
  };
}
```
It makes use of `cache.findAll` to get all of the specific `queryKeys`.

It would be nice to have something like a bit more supported out of the box though. There are good suggestions in this thread.

**Comment by shaimaHamila** [comment]
You can use setQueriesData to update multiple cache entries synchronously at the same time, you can filter the cache entries by partial keys to update the ones you need

I ran into the same issue, and seemed to solve it with a hook as follows:

`export const useCreateEquipment = () => {
  const queryClient = useQueryClient();
  return useMutation<ApiResponse<Partial<Equipment>>, Error, Partial<Equipment>>({
    mutationFn: async (newEquipment: Partial<Equipment>) => {
      const { data } = await api.post<ApiResponse<Partial<Equipment>>>("/equipment", newEquipment);
      return data;
    },
    onSuccess: (newEquipmentRes) => {
      notification.success({
        message: "Equipment created successfully",
      });

      const previousQueries = queryClient.getQueriesData<ApiResponse<Equipment[]>>({
        queryKey: ["equipment/fetchEquipments"],
      });
      // Optimistically update each query's data
      previousQueries.forEach(([queryKey, oldData]) => {
        if (oldData) {
          queryClient.setQueryData<ApiResponse<Partial<Equipment>[]>>(queryKey, {
            ...oldData,
            data: [...(oldData?.data || []), newEquipmentRes.data as Partial<Equipment>],
          });
        }
      });
    },
    onError: (error: any) => {
      notification.error({
        message: "Error",
        description: error.response?.data?.message || "Error",
      });
    },
  });
};`
